share/dkg/pedersen/dkg.go

```diff
@@ -1,143 +1,234 @@
+	return &Config{
+		Suite:     suite,
+		Longterm:  longterm,
+		NewNodes:  participants,
+		OldNodes:  participants,
+		Threshold: vss.MinimumT(len(participants)),
+	}
 }
 
-// Justification holds the Justification from a Dealer as well as the index of
-// the Dealer in question.
-type Justification struct {
-	// Index of the Dealer who answered with this Justification
-	Index uint32
-	// Justification issued from the Dealer
-	Justification *vss.Justification
+// NewReshareConfig returns a new config to use with DistKeyGenerator to run the
+// re-sharing protocols between the old nodes and the new nodes, i.e. the future
+// share holders. Share must be non-nil for previously enrolled nodes to
+// actively issue new shares. The public coefficients, pcoeffs, are needed in
+// order for participants in newNodes to be able to verify the validity of newly
+// received shares.
+func NewReshareConfig(suite Suite, longterm kyber.Scalar, oldNodes, newNodes []kyber.Point,
+	share *DistKeyShare, pcoeffs []kyber.Point) *Config {
+	return &Config{
+		Suite:        suite,
+		Longterm:     longterm,
+		OldNodes:     oldNodes,
+		NewNodes:     newNodes,
+		Share:        share,
+		PublicCoeffs: pcoeffs,
+		Threshold:    vss.MinimumT(len(newNodes)),
+	}
 }
 
 // DistKeyGenerator is the struct that runs the DKG protocol.
 type DistKeyGenerator struct {
+	// config driving the behavior of DistKeyGenerator
+	c     *Config
 	suite Suite
 
-	index uint32
-	long  kyber.Scalar
-	pub   kyber.Point
+	long   kyber.Scalar
+	pub    kyber.Point
+	dpub   *share.PubPoly
+	dealer *vss.Dealer
+	// verifiers indexed by dealer index
+	verifiers map[uint32]*vss.Verifier
 
-	participants []kyber.Point
+	// index in the old list of nodes
+	oidx int
+	// index in the new list of nodes
+	nidx int
+	// old threshold used in the previous DKG
+	oldT int
+	// new threshold to use in this round
+	newT int
+	// indicates whether we are in the re-sharing protocol or basic DKG
+	isResharing bool
+	// indicates whether we are able to issue shares or not
+	canIssue bool
+	// Indicates whether we are able to receive a new share or not
+	canReceive bool
+	// indicates whether the node holding the pub key is present in the new list
+	newPresent bool
+}
 
-	t int
+// NewDistKeyHandler takes a Config and returns a DistKeyGenerator that is able
+// to drive the DKG or resharing protocol.
+func NewDistKeyHandler(c *Config) (*DistKeyGenerator, error) {
+	if c.NewNodes == nil && c.OldNodes == nil {
+		return nil, errors.New("dkg: can't run with empty node list")
+	}
 
-	dealer    *vss.Dealer
-	verifiers map[uint32]*vss.Verifier
-}
+	var isResharing bool
+	if c.Share != nil || c.PublicCoeffs != nil {
+		isResharing = true
+	}
+	// canReceive is true by default since in the default DKG mode everyone
+	// participates
+	var canReceive = true
+	pub := c.Suite.Point().Mul(c.Longterm, nil)
+	oidx, oldPresent := findPub(c.OldNodes, pub)
```

Comment:
Can we use a `map[int]kyber.Point` for `OldNodes` and `NewNodes` in `Config` and avoid the `O(n)` overhead for the key search?

Reply:
The `O(n)` overhead will have to be done by the caller in that case. The caller most often has a simple list of pubic keys and not a map, so it has to create it first. I prefer less action for the caller.